src/ir/auth_logic/ast_construction.cc

```diff
@@ -0,0 +1,164 @@
+//-------------------------------------------------------------------------------
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+//-------------------------------------------------------------------------------
+
+#include <string>
+#include <vector>
+#include <utility>
+#include <stdexcept>
+#include <iostream>
+#include <memory>
+#include <variant>
+#include <optional>
+
+
+#include "antlr4-runtime.h"
+#include "absl/status/statusor.h"
+#include "src/common/utils/map_iter.h"
+#include "src/ir/auth_logic/ast.h"
+#include "src/ir/auth_logic/ast_construction.h"
+#include "src/policy/auth_logic/auth_logic_cc_generator_grammar.inc/AuthLogicLexer.h" 
+#include "src/policy/auth_logic/auth_logic_cc_generator_grammar.inc/AuthLogicParser.h"
+#include "src/policy/auth_logic/auth_logic_cc_generator_grammar.inc/AuthLogicVisitor.h"
+#include "src/ir/datalog/program.h"
+
+namespace raksha::ir::auth_logic {
+
+Principal Parser::ConstructPrincipal (AuthLogicParser::PrincipalContext* context) {
+    std::string name = context->ID()->getText();
+    return Principal(name);
+}
+
+datalog::Predicate Parser::ConstructPredicate (AuthLogicParser::PredicateContext* context ) {
+    std::string name = context->ID()->getText();
+    std::vector<std::string> args;
+    datalog::Sign sign;
+    
+    for (auto& statement : context->pred_arg()) {
+        args.push_back(std::move(statement->ID()->getText()));
+    }
+
+    sign = context->NEG() == nullptr ? datalog::Sign::kPositive : datalog::Sign::kNegated;
+    return datalog::Predicate(name, args, sign);
+
+}
+
+template<typename T, typename = std::enable_if_t<std::is_base_of<AuthLogicParser::VerbphraseContext, T>::value>>
+T* GetVerbPhraseContextAs(AuthLogicParser::VerbphraseContext* context) {
+     return dynamic_cast<T*>(context);
+}
+
+absl::StatusOr<BaseFact> Parser::ConstructVerbphrase(Principal left_principal, AuthLogicParser::VerbphraseContext* context) {
+    if (auto* ctx = GetVerbPhraseContextAs<AuthLogicParser::PredphraseContext>(context)) {
+
+        datalog::Predicate predicate = Parser::ConstructPredicate(ctx->predicate());   //PredicateContext
+        return BaseFact(Attribute(left_principal, datalog::Predicate(predicate)));
```

Comment:
`predicate` is already of type `Predicate`, why put this constructor call here? Also, you should probably use `std::move` on `predicate` or inline the initializer.